File: clients/typescript/test/batch.test.ts
import assert from "node:assert/strict";
import { after, test } from "node:test";

import { start, type Serving } from "./harness.js";
import {
  ge,
  lit,
  SlateError,
  type Batch,
  type Session,
  type Value,
  str,
  uint,
} from "../src/index.js";

/**
 * Several writes in one round trip, and the distinction that makes it safe.
 *
 * A batch is a network optimisation; a transaction is an atomicity guarantee.
 * They differ *only when something fails*, which is why `atomicity` has no
 * default here or on the wire. The two tests named `...carries on past a
 * failure` and `...undoes everything before the failure` are the same three
 * operations under the two guarantees, so the difference reads in one screen.
 */
const BATCH_TABLES = `
[[tables]]
name = "notes"
id = 43
columns = [
  { name = "id",   type = "u64" },
  { name = "body", type = "str" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["notes"]
actions = ["everything"]
`;

const servers: Serving[] = [];
after(() => {
  for (const s of servers) s.stop();
});

async function batched(): Promise<Session> {
  const server = await start(BATCH_TABLES);
  servers.push(server);
  return server.client().session();
}

function note(id: number): Value[] {
  return [uint(id), str("note")];
}

async function present(session: Session): Promise<number> {
  let seen = 0;
  for await (const _ of session.query({ table: "notes" })) seen++;
  return seen;
}

test("an empty batch is refused before it is sent", async () => {
  const session = await batched();
  await assert.rejects(
    () => session.batch({ atomicity: "independent", operations: [] }),
    (error: Error) => {
      assert.match(error.message, /at least one/);
      return true;
    },
  );
});

test("an independent batch applies every operation", async () => {
  const session = await batched();
  const batch: Batch = {
    atomicity: "independent",
    operations: [1, 2, 3, 4, 5].map((id) => ({
      kind: "insert" as const,
      table: "notes",
      rows: [note(id)],
    })),
  };
  const result = await session.batch(batch);

  assert.equal(result.outcomes.length, 5, "one outcome per operation");
  for (const one of result.outcomes) {
    assert.equal(one.error, undefined);
    assert.equal(one.written?.affected, 1n);
  }
  assert.notEqual(result.sequence, undefined);
  assert.equal(await present(session), 5);
});

test("an independent batch carries on past a failure", async () => {
  const session = await batched();
  await session.insert("notes", note(2));

  const result = await session.batch({
    atomicity: "independent",
    operations: [
      { kind: "insert", table: "notes", rows: [note(1)] },
      { kind: "insert", table: "notes", rows: [note(2)] }, // already there
      { kind: "insert", table: "notes", rows: [note(3)] },
    ],
  });

  assert.equal(result.outcomes[0]?.error, undefined);
  assert.equal(result.outcomes[2]?.error, undefined);
  const failed = result.outcomes[1]?.error;
  assert.ok(failed instanceof SlateError, `expected a SlateError, got ${failed}`);
  assert.equal(failed.kind, "already-exists");
  // The stable token survives the trip through a message body, which is the
  // half a lone call gets from its trailers.
  assert.equal(failed.reason, "DUPLICATE_PRIMARY_KEY");

  // 1 and 3 landed although 2 failed between them. That is independence, and
  // what a caller who wanted a transaction would be horrified by.
  assert.equal(await present(session), 3);
});

test("an atomic batch undoes everything before the failure", async () => {
  const session = await batched();
  await session.insert("notes", note(2));

  await assert.rejects(() =>
    session.batch({
      atomicity: "all-or-nothing",
      operations: [
        { kind: "insert", table: "notes", rows: [note(1)] },
        { kind: "insert", table: "notes", rows: [note(2)] }, // already there
        { kind: "insert", table: "notes", rows: [note(3)] },
      ],
    }),
  );

  // Only the seeded row: the same three operations as the test above.
  assert.equal(await present(session), 1);
});

test("an atomic batch reports no per-operation outcomes", async () => {
  const session = await batched();
  const result = await session.batch({
    atomicity: "all-or-nothing",
    operations: [1, 2, 3].map((id) => ({
      kind: "insert" as const,
      table: "notes",
      rows: [note(id)],
    })),
  });
  assert.deepEqual(result.outcomes, [], "they all happened; nothing to report");
  assert.notEqual(result.sequence, undefined);
  assert.equal(await present(session), 3);
});

test("a batch carries every kind of write", async () => {
  const session = await batched();
  await session.insert("notes", ...[1, 2, 3, 4, 5, 6].map(note));

  const result = await session.batch({
    atomicity: "independent",
    operations: [
      { kind: "insert", table: "notes", rows: [note(7)] },
      { kind: "insert", table: "notes", rows: [[uint(7), str("replaced")]], upsert: true },
      { kind: "delete", table: "notes", keys: [[uint(1)]] },
      {
        kind: "deleteWhere",
        write: { table: "notes", filter: ge(0, uint(6)), returning: true },
      },
      {
        kind: "updateWhere",
        write: { table: "notes", set: [{ column: 1, value: lit(str("touched")) }] },
      },
    ],
  });

  for (const [at, one] of result.outcomes.entries()) {
    assert.equal(one.error, undefined, `operation ${at} failed: ${one.error}`);
  }
  // The deleteWhere asked for its rows: 6 and 7.
  assert.equal(result.outcomes[3]?.written?.rows.length, 2);
  // The updateWhere did not ask, so it returns none.
  assert.equal(result.outcomes[4]?.written?.rows.length, 0);
  assert.equal(await present(session), 4);
});

test("an atomic batch joins an open transaction", async () => {
  const session = await batched();
  const tx = await session.begin();
  const result = await tx.batch({
    atomicity: "all-or-nothing",
    operations: [1, 2, 3].map((id) => ({
      kind: "insert" as const,
      table: "notes",
      rows: [note(id)],
    })),
  });
  // No sequence: the transaction has not committed, and the batch is not the
  // thing that commits it.
  assert.equal(result.sequence, undefined);
  await tx.rollback();
  assert.equal(await present(session), 0, "the rollback undid it");
});

test("an independent batch may not run inside a transaction", async () => {
  const session = await batched();
  const tx = await session.begin();
  await assert.rejects(() =>
    tx.batch({
      atomicity: "independent",
      operations: [{ kind: "insert", table: "notes", rows: [note(1)] }],
    }),
  );
  await tx.rollback();
});

File: clients/go/slate/batch_test.go
package slate_test

import (
	"errors"
	"strings"
	"testing"

	"github.com/howlerops/slate-orm/clients/go/slate"
)

const batchTables = `
[[tables]]
name = "notes"
id = 42
columns = [
  { name = "id",   type = "u64" },
  { name = "body", type = "str" },
]
primary_key = ["id"]

[[security.grants]]
role = "app"
tables = ["notes"]
actions = ["everything"]
`

func note(id uint64) []slate.Value {
	return []slate.Value{slate.Uint(id), slate.String("note")}
}

func batched(t *testing.T) *slate.Session {
	t.Helper()
	return start(t, batchTables).client(t).Session()
}

// present is every id in `notes`.
func present(t *testing.T, session *slate.Session) []uint64 {
	t.Helper()
	stream, err := session.Query(testContext(t), slate.Query{Table: "notes"})
	if err != nil {
		t.Fatalf("querying: %v", err)
	}
	defer stream.Close()
	out := []uint64{}
	for stream.Next() {
		row := stream.Row()
		id, ok := row[0].(slate.Uint)
		if !ok {
			t.Fatalf("the first column was not a u64: %v", row[0])
		}
		out = append(out, uint64(id))
	}
	if err := stream.Err(); err != nil {
		t.Fatalf("stream: %v", err)
	}
	return out
}

func TestBatchRefusesAnUnsetAtomicity(t *testing.T) {
	session := batched(t)
	// The zero value, which is what a caller who never chose would have.
	b := slate.NewBatch(slate.AtomicityUnset).Insert("notes", note(1))
	_, err := session.Batch(testContext(t), b)
	if err == nil {
		t.Fatal("a batch with no atomicity should be refused")
	}
	// The message must name both, because the caller's next move is to pick.
	for _, want := range []string{"Independent", "AllOrNothing"} {
		if !strings.Contains(err.Error(), want) {
			t.Fatalf("the refusal should name %q: %v", want, err)
		}
	}
	if got := present(t, session); len(got) != 0 {
		t.Fatalf("%d rows written on the way to being refused", len(got))
	}
}

func TestBatchRefusesAnEmptyList(t *testing.T) {
	session := batched(t)
	_, err := session.Batch(testContext(t), slate.NewBatch(slate.Independent))
	if err == nil {
		t.Fatal("an empty batch should be refused")
	}
}

func TestIndependentBatchAppliesEveryOperation(t *testing.T) {
	session := batched(t)
	b := slate.NewBatch(slate.Independent)
	for id := uint64(1); id <= 5; id++ {
		b.Insert("notes", note(id))
	}
	result, err := session.Batch(testContext(t), b)
	if err != nil {
		t.Fatalf("batching: %v", err)
	}
	if len(result.Outcomes) != 5 {
		t.Fatalf("%d outcomes, want one per operation", len(result.Outcomes))
	}
	for at, one := range result.Outcomes {
		if !one.OK() {
			t.Fatalf("operation %d failed: %v", at, one.Err)
		}
		if one.Written.Affected != 1 {
			t.Fatalf("operation %d affected %d, want 1", at, one.Written.Affected)
		}
	}
	if result.Sequence == nil {
		t.Fatal("the last commit's sequence should come back")
	}
	if got := len(present(t, session)); got != 5 {
		t.Fatalf("%d rows, want 5", got)
	}
}

// The difference, half one: a failure is a result, and the rest still land.
func TestIndependentBatchCarriesOnPastAFailure(t *testing.T) {
	session := batched(t)
	if _, err := session.Insert(testContext(t), "notes", note(2)); err != nil {
		t.Fatalf("seeding: %v", err)
	}

	b := slate.NewBatch(slate.Independent)
	b.Insert("notes", note(1))
	b.Insert("notes", note(2)) // already there
	b.Insert("notes", note(3))
	result, err := session.Batch(testContext(t), b)
	if err != nil {
		t.Fatalf("the batch itself should succeed: %v", err)
	}

	if result.Outcomes[0].OK() != true || result.Outcomes[2].OK() != true {
		t.Fatalf("the operations beside the failure should have landed: %+v", result.Outcomes)
	}
	failures := result.Failures()
	if len(failures) != 1 {
		t.Fatalf("%d failures, want 1", len(failures))
	}
	var failed *slate.Error
	if !errors.As(failures[0].Err, &failed) {
		t.Fatalf("a batch failure should be a *slate.Error: %v", failures[0].Err)
	}
	if failed.Kind != slate.KindAlreadyExists {
		t.Fatalf("kind = %v, want already-exists", failed.Kind)
	}
	// The stable token survives the trip through a message body, which is the
	// half a lone call gets from its trailers.
	if failed.Reason != "DUPLICATE_PRIMARY_KEY" {
		t.Fatalf("reason = %q, want DUPLICATE_PRIMARY_KEY", failed.Reason)
	}

	if got := len(present(t, session)); got != 3 {
		t.Fatalf("%d rows, want 3 — 1 and 3 land although 2 failed between them", got)
	}
}

// The difference, half two: the same three operations, and nothing lands.
func TestAtomicBatchUndoesEverythingBeforeTheFailure(t *testing.T) {
	session := batched(t)
	if _, err := session.Insert(testContext(t), "notes", note(2)); err != nil {
		t.Fatalf("seeding: %v", err)
	}

	b := slate.NewBatch(slate.AllOrNothing)
	b.Insert("notes", note(1))
	b.Insert("notes", note(2)) // already there
	b.Insert("notes", note(3))
	if _, err := session.Batch(testContext(t), b); err == nil {
		t.Fatal("an atomic batch should fail the call")
	}

	if got := present(t, session); len(got) != 1 || got[0] != 2 {
		t.Fatalf("rows = %v, want only the seeded 2", got)
	}
}

func TestAtomicBatchReportsNoPerOperationOutcomes(t *testing.T) {
	session := batched(t)
	b := slate.NewBatch(slate.AllOrNothing)
	for id := uint64(1); id <= 3; id++ {
		b.Insert("notes", note(id))
	}
	result, err := session.Batch(testContext(t), b)
	if err != nil {
		t.Fatalf("batching: %v", err)
	}
	if len(result.Outcomes) != 0 {
		t.Fatalf("%d outcomes; they all happened, so there is nothing to report", len(result.Outcomes))
	}
	if got := len(present(t, session)); got != 3 {
		t.Fatalf("%d rows, want 3", got)
	}
}

func TestBatchCarriesEveryKindOfWrite(t *testing.T) {
	session := batched(t)
	ctx := testContext(t)
	rows := make([][]slate.Value, 0, 6)
	for id := uint64(1); id <= 6; id++ {
		rows = append(rows, note(id))
	}
	if _, err := session.Insert(ctx, "notes", rows...); err != nil {
		t.Fatalf("seeding: %v", err)
	}

	b := slate.NewBatch(slate.Independent)
	b.Insert("notes", note(7))
	b.Upsert("notes", []slate.Value{slate.Uint(7), slate.String("replaced")})
	b.Delete("notes", []slate.Value{slate.Uint(1)})
	b.DeleteWhere(slate.DeleteWhere{
		Table:     "notes",
		Filter:    slate.Filter(slate.Ge(0, slate.Uint(6))),
		Returning: true,
	})
	b.UpdateWhere(slate.UpdateWhere{
		Table: "notes",
		Set:   []slate.Assignment{slate.Assign(1, slate.Lit(slate.String("touched")))},
	})
	result, err := session.Batch(ctx, b)
	if err != nil {
		t.Fatalf("batching: %v", err)
	}
	for at, one := range result.Outcomes {
		if !one.OK() {
			t.Fatalf("operation %d failed: %v", at, one.Err)
		}
	}
	// The delete_where asked for its rows: 6 and 7.
	if got := len(result.Outcomes[3].Written.Rows); got != 2 {
		t.Fatalf("%d returned rows, want 2 — returning should ride through a batch", got)
	}
	// The update_where did not ask, so it returns none.
	if got := len(result.Outcomes[4].Written.Rows); got != 0 {
		t.Fatalf("%d returned rows, want none", got)
	}
	if got := len(present(t, session)); got != 4 {
		t.Fatalf("%d rows, want 4", got)
	}
}

func TestAtomicBatchJoinsAnOpenTransaction(t *testing.T) {
	session := batched(t)
	ctx := testContext(t)
	tx, err := session.Begin(ctx)
	if err != nil {
		t.Fatalf("begin: %v", err)
	}
	b := slate.NewBatch(slate.AllOrNothing)
	for id := uint64(1); id <= 3; id++ {
		b.Insert("notes", note(id))
	}
	result, err := tx.Batch(ctx, b)
	if err != nil {
		t.Fatalf("batching inside a transaction: %v", err)
	}
	// No sequence: the transaction has not committed, and the batch is not the
	// thing that commits it.
	if result.Sequence != nil {
		t.Fatalf("sequence = %v, want none until commit", *result.Sequence)
	}
	if err := tx.Rollback(ctx); err != nil {
		t.Fatalf("rollback: %v", err)
	}
	if got := present(t, session); len(got) != 0 {
		t.Fatalf("%d rows after the rollback, want none", len(got))
	}
}

func TestIndependentBatchMayNotRunInsideATransaction(t *testing.T) {
	session := batched(t)
	ctx := testContext(t)
	tx, err := session.Begin(ctx)
	if err != nil {
		t.Fatalf("begin: %v", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()

	b := slate.NewBatch(slate.Independent).Insert("notes", note(1))
	if _, err := tx.Batch(ctx, b); err == nil {
		t.Fatal(`"independent operations that roll back together" should be refused`)
	}
}

// TestTheSchemaClaimRidesOnABatchedWrite: a misdeclared table is refused, and
// nothing lands.
//
// Found by a surviving mutation — dropping the claim from every operation
// changed no answer — and the same mutation survived in all three clients,
// which makes it a blind spot rather than an oversight. It matters as much
// here as anywhere: a batched insert whose claim is dropped is a write the
// server cannot check the shape of.
func TestTheSchemaClaimRidesOnABatchedWrite(t *testing.T) {
	server := start(t, batchTables)
	// `body` misdeclared as `text`: the same width and types, a different
	// name, which is exactly the drift a width check would miss.
	swapped := slate.TableDef{
		Name: "notes",
		Columns: []slate.ColumnDef{
			{Name: "id", Type: slate.TypeUint},
			{Name: "text", Type: slate.TypeString},
		},
		PrimaryKey: []string{"id"},
	}
	session := server.client(t).Declaring(slate.Schemas{"notes": swapped}).Session()

	b := slate.NewBatch(slate.Independent).Insert("notes", note(1))
	if _, err := session.Batch(testContext(t), b); !slate.IsKind(
		err, slate.KindInvalidRequest,
	) {
		t.Fatalf("a batch against a misdeclared table: err = %v, want KindInvalidRequest", err)
	}

	// And nothing was written on the way to being refused. Read through a
	// session that declares nothing, so the check above is not what hides it.
	plain := server.client(t).Session()
	if got := present(t, plain); len(got) != 0 {
		t.Fatalf("%d rows written by a refused batch", len(got))
	}
}
